File: scripts/Balanced Binary Tree/Balanced Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root: TreeNode) -> bool:

        # Initialize the result to True
        res = True

        # DFS through the tree
        def dfs(node, i):
            nonlocal res

            # If there isn't a node, return previous depth
            if not node:
                return i - 1

            # Check depths of the left and right subtrees
            left, right = dfs(node.left, i + 1), dfs(node.right, i + 1)

            # If they are more than 1 difference, save False to the result
            if abs(right - left) > 1:
                res = False

            # Return the max depth of both subtrees
            return max(left, right)

        dfs(root, 0)

        return res

File: scripts/algorithms/B/Balanced Binary Tree/Balanced Binary Tree.js
// Runtime: 112 ms (Top 53.69%) | Memory: 46.9 MB (Top 82.53%)
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 * this.val = (val===undefined ? 0 : val)
 * this.left = (left===undefined ? null : left)
 * this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {

    let result=true;
    function Dfs(root){
        if(!root)return 0;
        let left=Dfs(root.left)
        let right=Dfs(root.right)
        if(Math.abs(left-right)>1){result=false ;return;}
        return 1+Math.max(left,right)
    }
    Dfs(root)
    return result
};
